### PR TITLE
[WIP] add migration docs and release notes

### PR DESCRIPTION
**Pull request by ctb**
Add release notes and migration docs to the 4.0 doc update.

NOTE: PR into https://github.com/dib-lab/sourmash/pull/1283.

TODO:
- [x] add details about traverse-directory and lca summarize from [this comment](https://github.com/dib-lab/sourmash/issues/1069#issuecomment-676523543)
- [x] add details from [this comment](https://github.com/dib-lab/sourmash/issues/1069#issuecomment-682145116) about index and lca index command lines
- [x] add details from [this comment](https://github.com/dib-lab/sourmash/issues/1069#issuecomment-722410487) about MinHash changes - downsample and is_molecule type, in particular.

## Checklist

- [ ] Is it mergeable?
- [ ] `make test` Did it pass the tests?
- [ ] `make coverage` Is the new code covered?
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

File: doc/release-notes/sourmash-4.0.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+## Major changes for 4.0
+
+### New or changed behavior
+
+* default SBT storage is now .sbt.zip (#1174, #1170)
+* add `sourmash sketch` command for creating signatures (#1159)
+* protein ksizes in MinHash are now divided by 3, except in `sourmash compute` (#1277)
+* refactor MinHash API and implementation: add, iadd, merge, hashes, and max_hash (#1282, #1154, #1139, #1301)
+* add HyperLogLog implementation (#1223)
```
Comment (taylorreiter) [inline doc/release-notes/sourmash-4.0.md:25]: add HyperLogLog implementation of what :) 
May not be important to add that detail here, but also could be nice

File: doc/release-notes/sourmash-4.0.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+*  `lca summarize` no longer merges all signatures, and uses hash abundance by default (#1175)
+* `index `and `lca index` (#1186, #1222) now support `--from-file` and no longer require signature files on command line
+* `--traverse-directory` is now on by default for signature loading behavior (#1178)
+
+### Feature removal
+
+* remove Python 2.7 support (& end Python 2 compatibility) (#1145, #1144)
+* remove `lca gather` (#1307)
+* remove 10x support from `sourmash compute` (#1229)
+* remove 'dump' command (#1157)
```
Comment (taylorreiter) [inline doc/release-notes/sourmash-4.0.md]:
```suggestion
* remove `dump` command (#1157)
```

File: doc/release-notes/sourmash-4.0.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+### Feature/function deprecations 
+* deprecate `sourmash compute` (#1159)
+* deprecate `load_signatures`, `sourmash.load_one_signature`, `create_sbt_index`, and `load_sbt_index` (#1279, #1304)
+* deprecate import_csv in favor of new `sourmash sig import --csv` (#1281)
```
Comment (taylorreiter) [inline doc/release-notes/sourmash-4.0.md]:
```suggestion
* deprecate `import_csv` in favor of new `sourmash sig import --csv` (#1281)
```

File: doc/release-notes/sourmash-4.0.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+## Refactoring, improvements, and minor bug fixes:
+
+* accept file list in `sourmash sig cat` (#1236)
+* add unique_intersect_bp and gather_result_rank to gather CSV output (#1219)
+* remove deprecated minhash functions (#1149)
+* fix Rust panic error in signature creation (#1172)
+* cache nodes in SBT during search (#1161)
+* fix two bugs in gather --output-unassigned (#1156)
```
Comment (taylorreiter) [inline doc/release-notes/sourmash-4.0.md]:
```suggestion
* fix two bugs in gather `--output-unassigned` (#1156)
```

File: doc/support.md [hunk 5]
```diff
@@ -82,26 +86,60 @@ sourmash v3.x supports Python 2.7 as well as Python 3.x, through Python 3.8.
...
+
+Third, all sourmash commands that operate on signatures should now be able to directly read from lists of signatures in signature files, SBT databases, LCA databases, directories, and files containing lists of filenames (see [updated docs](command-line.md#advanced-command-line-usage)).
+
+Fourth, if you use `sourmash lca` commands, **`sourmash lca gather` has been removed**. In addition, there are some **changes in how `summarize` works**: it now uses abundances by default, and no longer combines all signatures before summarizing. Specify `--ignore-abundance` and combine your signatures using `sourmash sig merge` to recover the old behavior.
+
+Finally, **k-mer sizes have changed for amino acid sequences** in v4. if you use protein, Dayhoff, or HP signatures, we now interpret k-mer sizes differently on the command line. Briefly, k-mer sizes for protein/dayhoff/hp signatures are now the size of the k-mer in amino acid space, *not* the space of the k-mer in DNA space (as previously used). In practice this means that you need to divide all your k-mer sizes by 3 when working with k-mers in amino acid space!
```
Comment (taylorreiter) [inline doc/support.md]:
```suggestion
Finally, **k-mer sizes have changed for amino acid sequences** in v4. If you use protein, Dayhoff, or HP signatures, we now interpret k-mer sizes differently on the command line. Briefly, k-mer sizes for protein/dayhoff/hp signatures are now the size of the k-mer in amino acid space, *not* the space of the k-mer in DNA space (as previously used). In practice this means that you need to divide all your old k-mer sizes by 3 when working with k-mers in amino acid space!
```

File: doc/support.md [hunk 6]
```diff
@@ -82,26 +86,60 @@ sourmash v3.x supports Python 2.7 as well as Python 3.x, through Python 3.8.
...
+
+Note also that while `sourmash compute` still behaves the same way in v4.x as it did in sourmash 3.5.x, `sourmash sketch translate` and `sourmash sketch protein` both use the *new* approach to amino acid k-mer sizes, as do all of the the command line options for searching, manipulation, and display. Again, in practice this means that you need to divide all your k-mer sizes by 3 if they apply to amino acid k-mers.
```
Comment (taylorreiter) [inline doc/support.md]:
```suggestion
Note also that while `sourmash compute` still behaves the same way in v4.x as it did in sourmash 3.5.x, `sourmash sketch translate` and `sourmash sketch protein` both use the *new* approach to amino acid k-mer sizes, as do all of the the command line options for searching, manipulation, and display. Again, in practice this means that you need to divide all your old k-mer sizes by 3 if they apply to amino acid k-mers.
```

File: doc/support.md [hunk 7]
```diff
@@ -82,26 +86,60 @@ sourmash v3.x supports Python 2.7 as well as Python 3.x, through Python 3.8.
...
+### Python API
+
+First, all k-mer sizes for `protein`, `dayhoff`, and `hp` signatures have changed in the Python layer to be "correct", i.e., to be the size of the protein k-mer. Previously they were 3\*k, i.e. based on the size of the DNA k-mer from which the protein sequence would have been created.
 
-Prior to the release of sourmash v4, we are adding deprecation
-warnings and/or future warnings to all APIs and modules in sourmash
-v3.x that are being removed in v4.0. If you are using the Python API,
-we suggest you use the following procedure to migrate:
+Second, the `MinHash` class API has changed significantly!
+* `get_mins()` has been deprecated in favor of `.hashes`, which is a dictionary that contains abundances.
```
Comment (taylorreiter) [inline doc/support.md:136]: I simultaneously love this and hate how i'll have to fix all of my old code.

Comment (taylorreiter) [reply doc/support.md:136]: +1000 for being more intuitive though.

File: doc/support.md [hunk 8]
```diff
@@ -82,26 +86,60 @@ sourmash v3.x supports Python 2.7 as well as Python 3.x, through Python 3.8.
...
+* `merge` now just modifies `MinHash` objects in-place, and no longer returns the merged object; use `__iadd__` (`+=`) for the old behavior, or `__add__` (`+`) to create a new merged object.
+* `max_hash` has been deprecated in favor of `scaled`.
 
-* first, install the latest version of sourmash v3, which should be v3.5.0 or later.
-* then, turn on `DeprecationWarning`s in your code per [the warnings module documentation](https://docs.python.org/3/library/warnings.html#overriding-the-default-filter).
-* now, run python with the argument `-W error` to turn warnings into errors.
-* fix all errors!
-* finally, upgrade to sourmash v4.0.
+Third, `SourmashSignature` objects no longer have a `name()` method but instead a `name` property, which can be assigned to. This property is now `None` when no name has been assigned. Note that `str(sig)` should now be used to retrieve a display name, and should replace all previous uses of `sig.name()`.
```
Comment (taylorreiter) [inline doc/support.md:144]: this is great!

**Review by taylorreiter — COMMENTED** [review]
Great! Added some comments, compliments, and small suggestions.

File: doc/support.md (hunk 8 above)
Comment (ctb) [reply doc/support.md:144]: thanks!

File: doc/support.md (hunk 7 above)
Comment (ctb) [reply doc/support.md:136]: oh, well, you can totally keep using `get_mins()` until 5.0 removes it, in <checks date, makes predictions> 2025.

File: doc/release-notes/sourmash-4.0.md (hunk 1 above)
Comment (ctb) [reply doc/release-notes/sourmash-4.0.md:25]: I thought about this before, and decided not to. They can google or look at #1123 if they want - right now it's not particularly accessible and has no real impact on sourmash functionality, IIRC. But good catch :)

**Comment by ctb** [comment]
thank you very much @taylorreiter !!

**Review by taylorreiter — COMMENTED** [review]
This is great, I've added some comments, compliments, and suggestions.

**Comment by ctb** [comment]
@taylorreiter please see last set of changes here, b041bb8f.

@luizirber I'm happy to wait for your review if you like, just lmk.

**Comment by ctb** [comment]
thanks, fixed! gonna merge before luiz jumps in 😁

**Pull request merged**